### DISCUSSION TITLE
nonblock::Proxy and match signal

### DISCUSSION THREAD

**Discussion by pum-purum-pum-pum** [diwic/dbus-rs#401] [Q&A] [answered]
I'm trying to write an equivalent of the following code using `nonblock::Connection`:
```
use std::time::Duration;

use dbus::arg;
use dbus::blocking::Connection;
use dbus::channel::MatchingReceiver;
use dbus::message::MatchRule;
use dbus::Message;

#[derive(Debug)]
pub struct ManagerServicesChanged {
    pub changed: Vec<(dbus::Path<'static>, ::std::collections::HashMap<String, arg::Variant<Box<dyn arg::RefArg + 'static>>>)>,
    pub removed: Vec<dbus::Path<'static>>,
}

impl arg::AppendAll for ManagerServicesChanged {
    fn append(&self, i: &mut arg::IterAppend) {
        arg::RefArg::append(&self.changed, i);
        arg::RefArg::append(&self.removed, i);
    }
}

impl arg::ReadAll for ManagerServicesChanged {
    fn read(i: &mut arg::Iter) -> Result<Self, arg::TypeMismatchError> {
        Ok(ManagerServicesChanged { changed: i.read()?, removed: i.read()? })
    }
}

impl dbus::message::SignalArgs for ManagerServicesChanged {
    const NAME: &'static str = "ServicesChanged";
    const INTERFACE: &'static str = "net.connman.Manager";
}

fn main() {
    let conn = Connection::new_system().expect("D-Bus connection failed");
    conn.set_signal_match_mode(true);

    let rule = MatchRule::new();

    let proxy = conn.with_proxy("net.connman", "/", Duration::from_millis(5000));
    proxy
        .match_signal(|h: ManagerServicesChanged, _: &Connection, _: &Message| {
            dbg!();
            true
        })
        .unwrap();
    conn.start_receive(
        rule,
        Box::new(|msg, _| {
            handle_message(&msg);
            true
        }),
    );

    loop {
        conn.process(Duration::from_millis(1000)).unwrap();
    }
}

fn handle_message(msg: &Message) {
    println!("Got message: {:?}", msg);
}

```

What is not clear to me is how to match on the  `ServicesChanged` signal of the  `net.connman.Manager` interface(https://github.com/Ultimaker/connman/blob/master/doc/manager-api.txt#L211) with nonblock API.
Also `dbus-monitor --system` is printing the this message while the monitor example of `dbus-rs` (with the change to use system bus) -- doesn't.  I'm not sure why. I can only detect this message only if I write `proxy.match_signal` as in the example above

**Comment by diwic** [comment, marked as answer]
Have you looked at the [tokio_client](https://github.com/diwic/dbus-rs/blob/master/dbus-tokio/examples/tokio_client.rs) example? Maybe something like this (not tested) :

    use SignalArgs;
    let mr = ManagerServicesChanged::match_rule(None, "/");
    let incoming_signal = conn.add_match(mr).await?.cb(|_, h: ManagerServicesChanged| {
        dbg!(h);
        true
    });
    // Make sure incoming_signal is not dropped here! Leak it if necessary

**Reply by pum-purum-pum-pum** [reply]
I tried it again(that was the first thing I tried) and  it works this time. 
I was confused first time because I was doing it before spawning the task with the `resource.await` and it was failing with the timeout.
```
        runtime.spawn(async {
            let err = resource.await;
            // TODO should it it be log? Reconnection?
            panic!("Lost connection to D-Bus: {}", err);
        });
```
Thank you for the discussion!

**Comment by diwic** [comment]
> Also dbus-monitor --system is printing the this message while the monitor example of dbus-rs (with the change to use system bus) -- doesn't. I'm not sure why. I can only detect this message only if I write proxy.match_signal as in the example above

There is a stalled PR that could be about this here https://github.com/diwic/dbus-rs/pull/392 - will that help?
